File: library/builtins/primitives/evaluate.rs
use crate::*;
use wipple::*;

stored_closure!(pub struct EvaluateFn(&Env, &Stack) -> Result<Value>);

impl Primitive for EvaluateFn {}

#[ext(pub, name = ValueEvaluateExt)]
impl Value {
    #[track_caller]
    fn evaluate(&self, env: &Env, stack: &Stack) -> Result<Cow<Value>> {
        Ok(match self.get_if_present::<EvaluateFn>(env, stack)? {
            Some(evaluate) => Cow::Owned(evaluate.0(env, stack)?),
            None => Cow::Borrowed(self),
        })
    }
}

#[allow(clippy::unnecessary_wraps)]
pub(crate) fn setup(env: &Env, _stack: &Stack) -> Result<()> {
    env.set_variable("Evaluate", Value::of(Trait::of::<EvaluateFn>()));

    env.set_variable(
        "eval",
        Value::of(Function::new(|value, env, stack| {
            Ok(value
                .evaluate(env, stack)?
                .evaluate(env, stack)?
                .into_owned())
        })),
    );

    Ok(())
}

File: Library/Sources/Traits/Evaluate.swift
import Wipple

public struct Evaluate: Primitive {
    public typealias RawValue = (Env, Stack) throws -> Value

    public let rawValue: RawValue

    public init(_ rawValue: @escaping RawValue) {
        self.rawValue = rawValue
    }

    public func callAsFunction(_ env: Env, _ stack: Stack) throws -> Value {
        try self.rawValue(env, stack)
    }
}

public extension Value {
    func evaluate(_ env: Env, _ stack: Stack) throws -> Value {
        guard let evaluate = try self.getIfPresent(Evaluate.self, env, stack) else {
            return self
        }

        return try evaluate(env, stack)
    }
}

internal func setupEvaluate(_ env: Env, _ stack: Stack) throws {
    try env.setVariable("Evaluate", to: Value(Trait(Evaluate.self)), stack)

    try env.setVariable(
        "evaluate",
        to: Value(Function { value, env, stack in
            try value.evaluate(env, stack).evaluate(env, stack)
        }),
        stack
    )
}
